File: python/Facade/facade.py
from movie_element import Amplifier, DvdPlayer, PopcornPopper, Projecter, Screen, TheaterLights


class HomeTheterFacade:
    def __init__(self, amp: Amplifier, dvd: DvdPlayer, projector: Projecter, screen: Screen, lights: TheaterLights, popper: PopcornPopper) -> None:
        self.amp = amp
        self.dvd = dvd
        self.projector = projector
        self.screen = screen
        self.lights = lights
        self.popper = popper

    def watch_movie(self, movie: str):
        print("Get ready to watch a movie...")
        self.popper.on()
        self.popper.pop()
        self.lights.dim(10)
        self.screen.down()
        self.projector.on()
        self.amp.on()
        self.dvd.on()
        self.dvd.play(movie)

    def end_movie(self):
        print("Shutting movie theater down...")
        self.popper.off()
        self.lights.on()
        self.screen.up()
        self.projector.off()
        self.amp.off()
        self.dvd.stop()
        self.dvd.off()

File: c/Facade/facade.c
#include "facade.h"
void HTFacadeNew(HomeTheterFacade *self,
                 Amplifier *amp,
                 DvdPlayer *dvd,
                 Projector *projector,
                 Screen *screen,
                 TheaterLights *lights,
                 PopcornPopper *popper)
{
    self->amp = amp;
    self->dvd = dvd;
    self->projector = projector;
    self->screen = screen;
    self->lights = lights;
    self->popper = popper;
    self->WatchMovie = HTFacadeWatchMovie;
    self->EndMovie = HTFacadeEndMovie;
}
void HTFacadeWatchMovie(HomeTheterFacade *self, const char *movie)
{
    printf("Get ready to watch a movie...\n");
    self->popper->On(self->popper);
    self->popper->Pop(self->popper);
    self->lights->Dim(self->lights, 10);
    self->screen->Down(self->screen);
    self->projector->On(self->projector);
    self->amp->On(self->amp);
    self->dvd->On(self->dvd);
    self->dvd->Play(self->dvd, movie);
}
void HTFacadeEndMovie(HomeTheterFacade *self)
{
    printf("Shutting movie theater down...\n");
    self->popper->Off(self->popper);
    self->lights->On(self->lights);
    self->screen->Up(self->screen);
    self->projector->Off(self->projector);
    self->amp->Off(self->amp);
    self->dvd->Stop(self->dvd);
    self->dvd->Off(self->dvd);
}